src/modules/emote_autocomplete/components/Emotes.jsx

```diff
@@ -0,0 +1,120 @@
+import React, {useEffect, useRef, useState} from 'react';
+import {AutoSizer, List} from 'react-virtualized';
+import useChatInput from '../hooks/ChatInput.jsx';
+import styles from './Emotes.module.css';
+import useRowNavigation from '../hooks/RowKeyboardNavigation.jsx';
+import keyCodes from '../../../utils/keycodes.js';
+import renderRow from './EmoteRow.jsx';
+import EmotesHeader from './EmotesHeader.jsx';
+import emoteMenuViewStore from '../../../common/stores/emote-menu-view-store.js';
+
+const MAX_EMOTES_SHOWN = 8;
+const EMOTE_ROW_HEIGHT = 32;
+const BOTTOM_OFFSET = 120; // height of everything below the chat window
+
+function calcMaxHeight(length) {
+  let currentHeight = length * EMOTE_ROW_HEIGHT;
+
+  if (length > MAX_EMOTES_SHOWN) {
+    currentHeight = MAX_EMOTES_SHOWN * EMOTE_ROW_HEIGHT;
+  }
+
+  if (currentHeight > window.innerHeight - BOTTOM_OFFSET) {
+    currentHeight = window.innerHeight - BOTTOM_OFFSET;
+  }
+
+  return currentHeight;
+}
+
+let navigationCallback;
+function setNavigationCallback(callback) {
+  navigationCallback = callback;
+}
+
+export default function Emotes({chatInputElement, repositionPopover, onComplete, getAutocomplete}) {
+  const [emotes, setEmotes] = useChatInput(chatInputElement, getAutocomplete);
+  const [selected, setSelected] = useRowNavigation(setNavigationCallback, emotes.length);
+  const [open, setOpen] = useState(true);
+  const localRef = useRef(null);
+
+  useEffect(() => repositionPopover(), [emotes.length]);
+
+  function handleAutocomplete(emote) {
+    setEmotes([]);
+    emoteMenuViewStore.trackHistory(emotes);
+    onComplete(emote);
+  }
+
+  useEffect(() => {
+    function handleOutsideClick(event) {
+      if (localRef.current == null) {
+        return;
+      }
+
+      setOpen(localRef.current.contains(event.target));
+    }
+
+    window.addEventListener('mousedown', handleOutsideClick);
+
+    return () => {
+      window.removeEventListener('mousedown', handleOutsideClick);
+    };
+  }, [localRef.current, setOpen]);
+
+  useEffect(() => {
+    function keydownCallback(event) {
+      if (getAutocomplete() == null) {
+        setEmotes([]);
+        return;
+      }
+
+      setOpen(true);
+
+      if (event.key === keyCodes.Enter || event.key === keyCodes.Tab) {
+        event.stopPropagation();
+        event.preventDefault();
+        handleAutocomplete(emotes[selected]);
+        return;
+      }
+
+      navigationCallback(event);
+    }
+
+    window.addEventListener('keydown', keydownCallback, true);
+
+    return () => {
+      window.removeEventListener('keydown', keydownCallback, true);
+    };
+  }, [emotes, selected]);
+
+  if (!open || emotes.length === 0) {
+    return null;
+  }
+
+  return (
+    <div className={styles.emotes} ref={localRef}>
+      <EmotesHeader getAutocomplete={getAutocomplete} />
+      <AutoSizer disableHeight>
```

Comment:
we do we need this? could we show the first n results instead? as a user types more characters the list becomes more filtered, which gives them choices closer to what they want. we do not want people to browse the entire emote list here, but rather for them to press an arrow key or tab no more than a few times to get where they need to go